### ISSUE TITLE
LUTRAM write occurs on rising edge regardless of IS_WCLK_INVERTED property

### ISSUE THREAD

**Issue by hansemro** [openXC7/nextpnr-xilinx#33]
## Issue Description

Issue discovered in https://github.com/openXC7/nextpnr-xilinx/issues/20#issuecomment-1942858331

When IS_WCLK_INVERTED property is asserted, we expect the write to LUTRAM to occur on the falling edge. However, this does not happen as the placer and fasm writer ignores the property for LUTRAMs, and, therefore, treats as if all writes should occur on rising edge. This may lead to functional errors in designs that require writes on negative edge.

## Steps to Reproduce and Test

To distinguish between a write occuring on a positive edge and negative edge of clock, I created a simple test with two LUTRAMs sensitive to different edges of clock. Output of posedge LUTRAM and negedge LUTRAM drives led_o[0] and led_o[1], respectively. A single BSCANE2 primitive will be used to manually drive WCLK+DI+WE pins of the LUTRAM via JTAG.

1. Clone primitive-tests repo fork with `xc7-lutram-negedge-write-demo` branch

```
git clone https://github.com/hansemro/primitive-tests.git -b xc7-lutram-negedge-write-demo
```

2. Navigate to primitive-tests/lutram-tests/negedge-write-test and build the test with the openXC7 toolchain:

> [!NOTE]
> If targeting an FPGA board other than SQRL Acorn CLE215+, then provide an XDC file with
> 3 active-low LEDs mapped to led_o[2:0] and update `BOARD`/`FAMILY`/`PART`/`JTAG_CABLE`/`XDC`
> target parameters in the Makefile.

```
$ cd primitive-tests/lutram-tests/negedge-write-test
$ nix develop github:openXC7/toolchain-nix
[nix(openXC7)] make
```

3. Check that `CLKINV` bit is not set anywhere inside `top.fasm` (a sign that the test will fail). Proceed with the following steps to test.

4. Load the bitstream onto the FPGA:

```
make program
```

5. With LUTRAM initialized to 0, write a 1 on positive edge with the following openocd command:

> [!NOTE]
> Use an appropriate interface adapter script (usually located in `/usr/share/openocd/scripts/interface/`).
>
> Digilent HS2 support is provided with `digilent-hs2.cfg` in case `interface/ftdi/digilent-hs2.cfg` does not work.

> [!NOTE]
> `setup.cfg` defines a function `set_lutram <we> <data> <clk>` to manually set LUTRAM input pins.
>
> To avoid setup/hold violations, we should not change clk and we/data together.

```
openocd -f interface/ADAPTER.cfg -f setup.cfg \
    -c "set_lutram 0 0 0" \
    -c "set_lutram 0 0 0" \
    -c "set_lutram 1 1 0" \
    -c "set_lutram 1 1 1" \
    -c "shutdown"
```

If the design works correctly, only the LED (led_o[0]) assigned to the posedge LUTRAM should be lit after a positive-edge write. If LEDs led_o[1:0] light up together, then the design was incorrectly implemented. (LED at led_o[2] should dim on positive-edge write since the outputs should differ).

Anyhow, to perform a write on the negative edge, we would run the following afterwards:

```
openocd -f interface/ADAPTER.cfg -f setup.cfg \
    -c "set_lutram 1 1 0"
    -c "shutdown"
```

6. (Optional) Compare positive-edge (and negative-edge) write behavior against bitstream built with Vivado bitstream:

```
make vivadoclean top.vivado.bit
make BITSTREAM=top.vivado.bit program
# initial positive-edge write
openocd -f interface/ADAPTER.cfg -f setup.cfg \
    -c "set_lutram 0 0 0" \
    -c "set_lutram 0 0 0" \
    -c "set_lutram 1 1 0" \
    -c "set_lutram 1 1 1" \
    -c "shutdown"
# negative-edge write
openocd -f interface/ADAPTER.cfg -f setup.cfg \
    -c "set_lutram 1 1 0"
    -c "shutdown"
```

Design implemented with Vivado should have the two LUTRAM perform writes at different edges of the clock.

## Other Considerations

LUTRAM and FFs share clock from CLKINV routing BEL, so all synchronous elements within the SLICE site must have the same IS_*CLK_INVERTED property.

**Comment by hansfbaier** [comment]
Do you know how to disassemble a bitstream with bit2fasm?

**Comment by hansemro** [comment]
> Do you know how to disassemble a bitstream with bit2fasm?

Yes, here:

```
bit2fasm --db-root ${PRJXRAY_DB_DIR}/${FAMILY}/ --part ${PART} top.bit > top.fasm
```

**Comment by hansfbaier** [comment]
Not sure if I am comparing the right two parts here, but the CLKINV/NOCLKINV bits probably make the difference
![20240701_05h50m33s_grim](https://github.com/openXC7/nextpnr-xilinx/assets/148607/dfc739b9-a095-42f7-9e51-237bccfbba7e)
left is openXC7, right is vivado

**Comment by hansfbaier** [comment]
See `write_luts_config` and `write_ffs_config` in `fasm.cc`

**Comment by hansfbaier** [comment]
It might be a case for LUTRAMs is missing here:
![20240701_05h58m08s_grim](https://github.com/openXC7/nextpnr-xilinx/assets/148607/4202f5bc-055b-4a42-a5e2-7a6e6e5da8cd)

**Comment by hansemro** [comment]
@hansfbaier Yes, LUTRAM is currently not being handled in the fasm writer. Since LUTRAMs and FFs share the CLKINV BEL, I am thinking that the CLKINV bit should not be set inside `write_ffs_config` but inside a new function that handles both LUTRAMs and FFs. This new function would also be able to do some final DRC to check that the clock inversion property value matches for all used elements within the SLICE (half tile).

**Comment by hansemro** [comment]
There is a rule in `Arch::xc7_logic_tile_valid` to check that the FFs have the same inversion property:
https://github.com/openXC7/nextpnr-xilinx/blob/30ba8fbe77ace38e2e4ff08e9251d1e0d203c11d/xilinx/arch_place.cc#L557-L558

We may want to define a new rule to compare LUTRAM's clkinv and FF's clkinv.

**Comment by lehaifeng000** [comment]
good, I would join this work next week

**Comment by hansemro** [comment]
@lehaifeng000 @hansfbaier

Sounds good. I am still in the process of figuring out how to fully test whether the CLB placement rules work to catch problematic physical implementations. I suppose this is a good time to ask if anyone can reliably place LUTRAMs and FFs in the same CLB slice since I am failing to do this task with openxc7 flow.

**Comment by hansfbaier** [comment]
@hansemro How do you try to place them? xdc file, LOC annotations or other means?

**Comment by lehaifeng000** [comment]
In the place phase, is there any rule to avoid ff with different clkinv inside a slice?

**Comment by hansemro** [comment]
> @hansemro How do you try to place them? xdc file, LOC annotations or other means?

This may be an impossible task for the moment since we do not yet support manual BEL placement yet with LOC and BEL attributes. I was looking into setting CONSTR_{X,Y,Z} attributes in the post-pack json, but I am not entirely sure this is where we should be looking. Sorry for giving a rather loaded task; I don't know enough about nextpnr to give proper instructions.

**Comment by lehaifeng000** [comment]
> @hansemro How do you try to place them? xdc file, LOC annotations or other means?
@hansfbaier @hansemro 
After reviewing the code for "place", I believe a potential optimization could be implemented by modifying the "overused" condition within the "CutSpreader" function.

**Comment by hansfbaier** [comment]
@hansemro The primitives-fixes branch contains changes which seem to fix this issue.
Can you confirm that? What do your tests say?

**Comment by gHashTag** [comment]
@hansfbaier — answering your outstanding question directly: **yes, `primitives-fixes` contains the mechanism this issue needs.** Evidence below, plus the reason it has not been confirmable.

## The branch does address it

`stable-backports` derives the half-slice clock polarity from **flip-flops only** (`fasm.cc`, `is_clkinv` is set inside the FF loop via `SET_CHECK`). A half-slice holding a LUTRAM with an inverted write clock and no FF has nothing to carry the polarity, so it falls through as `NOCLKINV` — and there is exactly one bit for the whole half-slice:

```
CLBLM_L.SLICEM_X0.CLKINV   01_51
CLBLM_L.SLICEM_X0.NOCLKINV !01_51
```

`primitives-fixes` replaces that with a per-tile `ClkStatus` established during packing rather than inferred at FASM time, and `pack_dram.cc` sets it on the RAM cells themselves:

```
+ base->attrs[ctx->id("CLK_STATUS")] = Property("CLKINV");
+ ram_lut->attrs[ctx->id("CLK_STATUS")] = Property("CLKINV");
```

That is the right shape — the polarity now comes from whichever element actually knows it, instead of from a element type that may not be present.

## Why nobody has been able to confirm it

Two independent reasons, both worth knowing:

**1. The reproducer needs BSCANE2, which cannot currently be built to a bitstream.** The test in this issue drives WCLK/DI/WE through a `BSCANE2`. On artix7 `prjxray-db` documents only the four `BSCAN.JTAG_CHAIN_*` site bits and **zero** `CFG_CENTER_LOGIC_OUTS` pips, so the primitive places and routes and then dies in `fasm2frames` with a `FasmLookupError`. I filed that separately as #126. So the confirmation you asked for was blocked on something unrelated to this issue.

**2. The branch is 163 commits behind.** `origin/primitives-fixes` is 4 ahead of `stable-backports` and 163 behind it. That, rather than any doubt about the fix, looks like what has kept it out.

## What I checked, and what I am not claiming

I went looking for this from the other end — a sweep for half-written inverter pairs turned up the FF-only `is_clkinv` derivation independently, and I started writing the same fix before finding yours. I dropped mine: a narrower duplicate of a better change already on a branch is not worth reviewing.

I could **not** produce a failing case on `stable-backports`, and I want to be precise about that rather than imply I confirmed the bug:

* the inferred path (`reg mem[0:31]; always @(negedge clk) ...`) gives a `RAM32M` with `IS_WCLK_INVERTED` unset and a **separate `INV` cell** — the inversion lands in the fabric, so `NOCLKINV` is correct there and no bug is visible;
* the direct path (`RAM32X1S #(.IS_WCLK_INVERTED(1'b1))`) fails earlier with `Cannot pack unsupported primitive: RAM32X1S` — and `RAM32X1S` is one of the primitives `primitives-fixes` implements.

So on current master the bug may well be **unreachable**, which would explain the difficulty of demonstrating it, and would mean this issue is really "waiting on primitives-fixes" rather than "waiting on a diagnosis".

## Suggestion

The valuable work here is not another patch, it is getting that branch landable: 1385 lines implementing `RAM32X1S/RAM64X1S/RAM128X1D/RAM256X1S/RAMS32/RAMS64E/CFGLUT5/LDCE/LDPE/FIFO18E1/FIFO36E1/DNA` and fixing `IDDR/ODDR/BUFGMUX_1/BUFGCE_1/ICAP` — which touches this issue, #20, #75 and probably others.

If you want it rebased I am happy to take that on, but I would rather ask first than push 163 commits of conflict resolution at you unasked. Tell me whether you want it as one rebase or split per primitive, and I will start.

**Comment by gHashTag** [comment]
Follow-up with an actual number, since "163 commits behind" sounds worse than it is.

I merged `origin/stable-backports` into `origin/primitives-fixes` in a scratch worktree to measure the damage:

```
$ git merge origin/stable-backports --no-commit --no-ff
$ git diff --name-only --diff-filter=U
xilinx/arch_place.cc          1 hunk
xilinx/constids.inc           1
xilinx/fasm.cc                2
xilinx/pack.cc                3
xilinx/pack_clocking_xc7.cc   2
xilinx/pack_dram.cc           1
xilinx/pack_io_xc7.cc         1
xilinx/xdc.cc                 1
                             --
                             12 conflict hunks, 8 files
```

**Twelve hunks** for 1385 lines of work across 163 commits of drift. That is an afternoon, not a rewrite — the branch and master have largely been growing in different places.

Which reframes it: what is keeping `RAM32X1S/RAM64X1S/RAM128X1D/RAM256X1S/RAMS32/RAMS64E/CFGLUT5/LDCE/LDPE/FIFO18E1/FIFO36E1/DNA` and the `IDDR/ODDR/BUFGMUX_1/BUFGCE_1/ICAP` fixes out of tree is not the size of the merge.

The offer stands and I would rather do this than write more one-line patches around the edges. What I would want from you first:

1. **Rebase or merge?** A merge commit keeps @hansemro's history readable; a rebase gives a clean series. Your repo, your preference.
2. **One PR or split per primitive?** Split is much easier to review and lets the safe ones land while the arguable ones are discussed, but it is a lot more PRs.
3. Anything on that branch you have since decided against, so I do not carry it forward.

Caveat I want to be upfront about, given how tonight went: I can resolve the conflicts and get it to compile, and I can run the FASM-level regression suite over it. I **cannot** validate most of those primitives — I have no Vivado and no vendor bitstream containing a `FIFO18E1` or a `CFGLUT5`, so "it builds and emits plausible FASM" is the ceiling of what I can promise. Three database-derived changes went through the tool tonight and two behaved the opposite of how the database read, so I have stopped treating "the sources say so" as sufficient.

**Comment by hansfbaier** [comment]
@gHashTag I definitely would prefer a rebase

**Comment by hansfbaier** [comment]
@gHashTag  PR granularity is not important. But we definitely need one commit per primitive to do bisects if things break.

**Comment by hansfbaier** [comment]
3. @gHashTag  There is not anything in this branch I actually have decided against, I just was hesitant because I did not have the time and energy to understand the involved issues, and there were no automated tests.

**Comment by gHashTag** [comment]
Partitioned `primitives-fixes` by primitive for the rebase you asked for. **The partition turned up something that changes the plan: a straight replay would regress four primitives you have already merged.**

## `pack_dram.cc` overlaps with `9ec2b030`

Your `stable-backports` already handles all four of these — `pack_dram.cc:373` and `:426`:

```cpp
} else if (cs.memtype == ctx->id("RAM128X1D") || cs.memtype == ctx->id("RAM256X1D")) {
} else if (cs.memtype == ctx->id("RAM128X1S") || cs.memtype == ctx->id("RAM256X1S")) {
```

from `9ec2b030` ("RAM128/256X1S distributed-RAM packing + INT-tile constant-net bridge"), which post-dates the branch. So:

* **RAM256X1S** — branch version is a duplicate. Drop.
* **RAM128X1S** — likewise, *except* for a genuine 2-line delta: `dt.abits <= (7 - (dt.rports > 0))` in both address loops, which makes it read scalar `A0..A6` per the unisim instead of `A[6:0]`. Worth keeping on its own; the rest is not.
* **RAM128X1D** — the branch **deletes the `m256` path**. `RAM256X1D` then has no `memtype ==` handler at all while staying registered in `dram_types` and `pins.cc:154`, so it falls off the end of the if/else chain silently. Verified: one mention of `RAM256X1D` left in the branch's `pack_dram.cc`, and no handler.

The branch versions also hardcode the muxf `z` base, losing upstream's `ctx->xc7 ? 2 : 6` split.

The same shape as the IDDR case I mentioned earlier: the branch's claim was true when written and has since been overtaken.

## Two more things the partition found

**The `log_error` catch-all must not be deleted wholesale.** The branch removes the whole clause; only the `RAM32X1S` and `RAM64X1S` disjuncts are obsolete. `RAMS32`/`RAMD32`/`RAMS64E`/`RAMD64E` must keep erroring — dropping the clause is *why* `RAM256X1D` would fail silently rather than loudly.

**Three hunks are a revert of your `b649145a`**, not primitive work — the `dram_site`/`pinned` handling in the RAM32M/RAM64M whole-SLICE loop. They exist only because the branch predates it. Dropping.

Also: the commit message advertises `RAM32X2S_1`, which exists nowhere in the branch. What is actually there is `RAM32X1S_1`.

## What the rebase will therefore contain

Genuinely new, one commit each: **RAM64X1S, RAM32X1S, RAM32X2S, RAM32X1S_1, RAM64X1S_1, RAMS32, RAMS64E, CFGLUT5, LDCE, LDPE, IBUF_IBUFDISABLE, DNA, FIFO18E1, FIFO36E1** — plus the RAM128X1S naming delta, and a small shared-prerequisite commit first (32-deep DRAM moving to the high LUT5 with `A6=VCC`, which every `RAM32X*` group depends on).

`BUFGMUX` stays in its own commit, flagged, pending your call on regymm's `343457cc`.

Ordering constraint worth knowing: `constids.inc` is positional and `known_id_count` is baked into the chipdb, so its hunk lands in whichever commit first needs the new ids, and **chipdbs need regenerating after the series**.

I will build the series next and open it as a fresh PR. Wanted you to see the drops first, since "rebase it" and "rebase it minus four primitives you already have" are different things and the second is the honest one.

**Comment by gHashTag** [comment]
## Retracting one line from my inventory above

In my 2026-08-11 partition comment I wrote of the `primitives-fixes` branch:

> **RAM128X1S** — likewise, *except* for a genuine 2-line delta: `dt.abits <= (7 - (dt.rports > 0))`
> in both address loops, which makes it read scalar `A0..A6` per the unisim instead of `A[6:0]`.
> **Worth keeping on its own**; the rest is not.

and carried it into the rebase plan as "plus the RAM128X1S naming delta".

**That is wrong, and I have measured it rather than reasoned about it.** Triple comparison
against a Vivado golden:

```
upstream                        WA7USED 0   WEMUX 1
abits <= (7 - (rports > 0))     WA7USED 1   WEMUX 0     <- the branch's delta
@cavearr's #144                 WA7USED ok  WEMUX.CE ok
```

The delta recovers the write address and **loses the write enable**. It is half a fix, and the
dangerous half: the RAM places and routes cleanly while reading from floating addresses and
never writing. That is worse than the unfixed version, because the unfixed version at least
fails in a way somebody notices.

The correct fix is scalar-first lookup with bus fallback — **#144**, which I ran on
xc7a200t before it merged.

**Action for whoever builds the rebase series:** drop "plus the RAM128X1S naming delta" from
the plan in my comment above. If you have already started from that list, skip that item.

Apologies for the noise. I put it in the plan on a source reading and only measured it after
@cavearr produced a golden that disagreed.

**Comment by hansfbaier** [comment]
Verified fixed at the bitstream level — closing, with one silicon caveat at the end.

The defect was that a LUT-RAM's `IS_WCLK_INVERTED` never reached the half-slice clock inverter. PR openXC7/nextpnr-xilinx#203 (`e35b57af`, *drive the half-slice CLKINV from the LUT-RAM clock inversion*, in `main` and in the 0.9.6 release) folds the LUT-RAM cells into the same `is_clkinv` decision as the flip-flops: `xilinx/fasm.cc` now tracks `mem_clkinv` and flags a memory/FF disagreement (`mem_and_ff_disagree`).

Measured on `xc7a35tcsg324-1` with two otherwise identical `RAM32X1D` designs, one with `IS_WCLK_INVERTED=1`:

    diff ram32x1d_plain/top.fasm ram32x1d_inv/top.fasm
    13c13
    < CLBLM_R_X41Y57.SLICEM_X0.NOCLKINV
    ---
    > CLBLM_R_X41Y57.SLICEM_X0.CLKINV

One bit, on the RAM's own half-slice — the parameter now reaches the shared slice inverter instead of being ignored (previously the two FASM files were byte-identical). The negedge case lands correctly too: the two edge-sensitive RAMs of the repro program opposite bits in different slices (`…X41Y64.SLICEM_X0.NOCLKINV` for the posedge RAM, `…X41Y58.SLICEM_X0.CLKINV` for the negedge one). The canonical regression case `demo-projects/regression/lutram-clkinv` (`RAM64X1S` with `IS_WCLK_INVERTED=1`) passes its `check.sh`: `ok: SLICEM_X0 CLKINV with a 64x1 LUT-RAM (no NOCLKINV, no SMALL, no WA7/WA8USED)`.

Silicon caveat: the runtime check in the report (BSCANE2 + JTAG, which LED lights on which edge) needs a board, and the BSCANE2 path still does not build to a bitstream on this tree (#126), so the falling-edge *write* was not confirmed on hardware — only the bitstream bit that selects the edge. If the board test still shows the wrong edge with that bit now set, reopen and I will chase the remaining piece.
